### PR TITLE
Added dependency upgrades for python3.12 support

### PR DESCRIPTION
**Pull request by shivakrishnaah**
This PR adds support for python 3.11 and 3.12 to mlserver.

As `alibi-explain` and `alibi-detect` runtimes required to be also upgraded to allow this support (and not yet released), we currently point to the master branches instead for installation.

Before a proper release of mlserver, we need to revert this change. This also required a proper release of the Alibi libraries.

The main changes were on TensorFlow and Pytorch upgrades to support this transition.

Also the CI now contains tests for python 3.9-3.12 which we officially support.

### CODE REVIEW

**Comment by shivakrishnaah** [comment]
# Pull Request
fixes #1926 

## Description
Upgraded most of the dependencies to support python3.12 version

## Changes Made
Upgraded the dependencies and the python docker images

## Checklist


- [X ] Code follows the project's style guidelines
- [ ] All tests related to the changes pass successfully
- [ ] Documentation is updated (if necessary)
- [ ] Code is reviewed by at least one other team member
- [ ] Any breaking changes are communicated and documented

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -67,6 +68,7 @@ jobs:
         python-version:
           - "3.9"
           - "3.10"
+          - "3.12"
```
Comment (sakoush) [inline .github/workflows/tests.yml:71]: add 3.11

File: .github/workflows/benchmark.yml [hunk 2]
```diff
@@ -15,10 +15,10 @@ jobs:
         scenario: ["inference-rest.js", "inference-grpc.js"]
     steps:
       - uses: actions/checkout@v4
-      - name: Set up Python 3.10
```
Comment (sakoush) [inline .github/workflows/benchmark.yml:18]: lets keep the CI at python `3.10` as it is related to the main docker python version, and just add support for 3.11 and 3.12 if thats possible.

File: .github/workflows/benchmark.yml [hunk 3]
```diff
@@ -15,10 +15,10 @@ jobs:
...
+      - name: Set up Python 3.12
         uses: actions/setup-python@v5
         with:
-          python-version: "3.10"
```
Comment (sakoush) [inline .github/workflows/benchmark.yml:21]: ditto

File: .github/workflows/licenses.yml [hunk 4]
```diff
@@ -12,10 +12,10 @@ jobs:
     if: github.repository == 'SeldonIO/MLServer' # Do not run this on forks.
     steps:
       - uses: actions/checkout@v4
-      - name: Set up Python 3.10
```
Comment (sakoush) [inline .github/workflows/licenses.yml:15]: ditto

File: .github/workflows/licenses.yml [hunk 5]
```diff
@@ -12,10 +12,10 @@ jobs:
...
+      - name: Set up Python 3.12
         uses: actions/setup-python@v5
         with:
-          python-version: "3.10"
```
Comment (sakoush) [inline .github/workflows/licenses.yml:18]: ditto

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -185,10 +185,10 @@ jobs:
       - uses: actions/checkout@v4
         with:
           ref: ${{ github.event.inputs.version }}
-      - name: Set up Python 3.10
```
Comment (sakoush) [inline .github/workflows/release.yml:188]: ditto

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -185,10 +185,10 @@ jobs:
...
+      - name: Set up Python 3.12
         uses: actions/setup-python@v5
         with:
-          python-version: "3.10"
```
Comment (sakoush) [inline .github/workflows/release.yml:191]: ditto

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -231,10 +231,10 @@ jobs:
       - uses: actions/checkout@v4
         with:
           ref: ${{ github.event.inputs.version }}
-      - name: Set up Python 3.10
```
Comment (sakoush) [inline .github/workflows/release.yml:234]: ditto

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -231,10 +231,10 @@ jobs:
...
+      - name: Set up Python 3.12
         uses: actions/setup-python@v5
         with:
-          python-version: "3.10"
```
Comment (sakoush) [inline .github/workflows/release.yml:237]: ditto

File: .github/workflows/tests.yml [hunk 10]
```diff
@@ -39,6 +39,7 @@ jobs:
...
```
Comment (sakoush) [inline .github/workflows/tests.yml:42]: add 3.11

File: .github/workflows/tests.yml [hunk 11]
```diff
@@ -169,6 +173,7 @@ jobs:
...
```
Comment (sakoush) [inline .github/workflows/tests.yml:176]: add 3.11

File: .readthedocs.yaml [hunk 12]
```diff
@@ -16,7 +16,7 @@ formats:
 build:
   os: ubuntu-22.04
   tools:
-    python: "3.10"
```
Comment (sakoush) [inline .readthedocs.yaml:19]: ditto

File: Dockerfile [hunk 13]
```diff
@@ -1,7 +1,7 @@
-FROM python:3.10-slim AS wheel-builder
```
Comment (sakoush) [inline Dockerfile:1]: lets leave the default mlserver docker image that we build as part of the CI at python 3.10, if we want a docker image for different python version then the options are:
- parameterise this `Dockerfile` to take the python version as argument.
- provide instructions on how to build a docker image with a different python version

In any case I think this should be done as a follow up PR perhaps.

File: docs-gb/examples/conda/environment.yml [hunk 14]
```diff
@@ -3,11 +3,11 @@ name: old-sklearn
 channels:
     - conda-forge
 dependencies:
-    - python == 3.8
```
Comment (sakoush) [inline docs-gb/examples/conda/environment.yml:6]: these are part of examples, perhaps leave them as they are?

File: docs/conf.py [hunk 15]
```diff
@@ -78,7 +78,7 @@
         "base_url": "https://mlserver.readthedocs.io",
         "color_primary": "teal",
         "color_accent": "light-blue",
-        "repo_url": "https://github.com/SeldonIO/MLServer/",
```
Comment (sakoush) [inline docs/conf.py:81]: revert

File: docs/examples/conda/environment.yml [hunk 16]
```diff
@@ -3,11 +3,11 @@ name: old-sklearn
 channels:
     - conda-forge
 dependencies:
-    - python == 3.8
```
Comment (sakoush) [inline docs/examples/conda/environment.yml:6]: ditto

File: mlserver/cli/constants.py [hunk 17]
```diff
@@ -11,7 +11,9 @@
 RUN conda config --add channels conda-forge && \\
     conda install conda-libmamba-solver==23.7.0 && \\
     conda config --set solver libmamba && \\
-    conda install conda-pack
+    conda install conda-pack && \\
+    conda install gcc_linux-aarch64 && \\
+    conda install gxx_linux-aarch64 && source activate
```
Comment (sakoush) [inline mlserver/cli/constants.py:16]: why are these required?

File: poetry.lock [hunk 18]
```diff
@@ -1,4 +1,4 @@
-# This file is automatically @generated by Poetry 1.8.3 and should not be changed by hand.
+# This file is automatically @generated by Poetry 1.8.4 and should not be changed by hand.
```
Comment (sakoush) [inline poetry.lock:1]: is there a reason why the lock file has to change almost completely? There are internal libraries that have been downgraded (e.g. `alibi` and `alibi-detect`) that at least should be reverted.

File: pyproject.toml [hunk 19]
```diff
@@ -80,6 +80,7 @@ tox = "4.16.0"
 
 [tool.poetry.group.dev.dependencies]
 datamodel-code-generator = "0.26.0"
+pydantic = "*"
```
Comment (sakoush) [inline pyproject.toml:83]: is this redundant as above `pydantic` rule is defined?

File: pyproject.toml [hunk 20]
```diff
@@ -111,7 +112,8 @@ filelock = "^3.13.1"
 httpx-sse = ">=0.3.1,<0.5.0"
 
 [tool.poetry.group.docker.dependencies]
-tensorflow = "<2.15"
+#tensorflow-macos = { version = "^2.15.0", markers = "platform_system=='Darwin' and platform_machine=='arm64'" }
+tensorflow = ">=2.16"
```
Comment (sakoush) [inline pyproject.toml:116]: why are these required?

File: runtimes/alibi-detect/pyproject.toml [hunk 21]
```diff
@@ -8,13 +8,15 @@ readme = "README.md"
 packages = [{include = "mlserver_alibi_detect"}]
 
 [tool.poetry.dependencies]
-python = ">=3.9,<3.12"
+python = ">=3.9,<3.13"
 mlserver = "*"
-alibi-detect = {extras = ["tensorflow"], version = "*"}
+alibi-detect = "0.12.0"
+tensorflow = ">=2.16"
```
Comment (sakoush) [inline runtimes/alibi-detect/pyproject.toml:14]: why are these required?

File: runtimes/alibi-detect/pyproject.toml [hunk 22]
```diff
@@ -8,13 +8,15 @@ readme = "README.md"
...
 
 [tool.poetry.group.dev.dependencies]
 mlserver = {path = "../..", develop = true}
-tensorflow = "<2.15"
+#tensorflow-macos = { version = "^2.15.0", markers = "platform_system=='Darwin' and platform_machine=='arm64'" }
+tensorflow = ">=2.16"
```
Comment (sakoush) [inline runtimes/alibi-detect/pyproject.toml:19]: why are these required?

File: runtimes/alibi-explain/pyproject.toml [hunk 23]
```diff
@@ -8,20 +8,21 @@ readme = "README.md"
...
 [tool.poetry.dependencies]
-python = ">=3.9,<3.12"
+python = ">=3.9,<3.13"
 mlserver = "*"
 mlserver_sklearn = "*"
 mlserver_xgboost = "*"
 mlserver_lightgbm = "*"
 orjson = "*"
-alibi = {extras = ["shap", "tensorflow"], version = "*"}
+alibi = "0.9.6"
```
Comment (sakoush) [inline runtimes/alibi-explain/pyproject.toml:17]: why is this required? this is an old version.

File: runtimes/alibi-explain/pyproject.toml [hunk 24]
```diff
@@ -8,20 +8,21 @@ readme = "README.md"
...
 
 [tool.poetry.group.dev.dependencies]
 mlserver = {path = "../..", develop = true}
 mlserver_sklearn = {path = "../../runtimes/sklearn", develop = true}
 mlserver_xgboost = {path = "../../runtimes/xgboost", develop = true}
 mlserver_lightgbm = {path = "../../runtimes/lightgbm", develop = true}
-tensorflow = "<2.15"
+#tensorflow-macos = { version = "^2.15.0", markers = "platform_system=='Darwin' and platform_machine=='arm64'" }
+tensorflow = ">=2.16"
```
Comment (sakoush) [inline runtimes/alibi-explain/pyproject.toml:25]: ditto

File: tests/test_env.py [hunk 25]
```diff
@@ -47,7 +47,7 @@ def test_sys_path_empty(env: Environment):
 
 @pytest.mark.parametrize(
     "folder_name",
-    ["python3.9", "python37", "python3.11", ""],
+    ["python3.9", "python3.11", "python3.12"],
```
Comment (sakoush) [inline tests/test_env.py:50]: keep the extra empty python version `""`

File: tests/testdata/environment.yml [hunk 26]
```diff
@@ -2,7 +2,9 @@ name: custom-runtime-environment
 channels:
   - conda-forge
 dependencies:
-  - python == 3.9
-  - scikit-learn == 1.0.2
+  - python == 3.12
```
Comment (sakoush) [inline tests/testdata/environment.yml:5]: ditto

File: tox.ini [hunk 27]
```diff
@@ -37,7 +37,7 @@ commands =
         {toxinidir}/tests/grpc/test_interceptor.py
 set_env =
     GITHUB_SERVER_URL = {env:GITHUB_SERVER_URL:https\://github.com}
-    GITHUB_REPOSITORY = {env:GITHUB_REPOSITORY:SeldonIO/MLServer}
+    GITHUB_REPOSITORY = {env:GITHUB_REPOSITORY:shivakrishnaah/MLServer}
```
Comment (sakoush) [inline tox.ini:40]: revert

File: tox.ini [hunk 28]
```diff
@@ -65,7 +65,7 @@ commands =
         {toxinidir}/tests/parallel
 set_env =
     GITHUB_SERVER_URL = {env:GITHUB_SERVER_URL:https\://github.com}
-    GITHUB_REPOSITORY = {env:GITHUB_REPOSITORY:SeldonIO/MLServer}
+    GITHUB_REPOSITORY = {env:GITHUB_REPOSITORY:shivakrishnaah/MLServer}
```
Comment (sakoush) [inline tox.ini:68]: revert

**Review by sakoush — CHANGES_REQUESTED** [review]
@shivakrishnaah many thanks for your great contribution to add support for python 3.12.

To keep the changes for this PR as minimal as possible, lets _just_ add support for python 3.12 and keep the default version of mlserver at python 3.10 for now. We cannot upgrade everything to python 3.12 at the moment as this will break other workflows that we have.

So lets revert the changes to the CI / dockerfile.

I have triggered the CI anyway so that we can look at the tests status.

**Comment by sakoush** [comment]
@shivakrishnaah Many thanks for your recent changes. There are some issues with the CI (check [here](https://github.com/SeldonIO/MLServer/actions/runs/11862265518/job/33081687689?pr=1951) as an example):

`pyproject.toml changed significantly since poetry.lock was last generated. Run `poetry lock [--no-update]` to fix the lock file.`


The first thing that needs to be fixed perhaps on your PR is to run `make lock` at the top level to refresh poetry lock files for mlserver.

There are a few things to note on top:
- Torch min version required for py 3.12 support is py 2.4, check [here](https://github.com/pytorch/pytorch/releases/tag/v2.4.0). This needs to be updated.
- Some tests are fail in my local env. They will need to be addressed.
- Some internal libraries (e.g. `alibi` and `alibi-detect` pin their TF dependencies to be <2.15 and therefore we will also have to upgrade these internal libraries before accepting this change (we will upgrade these libraries ourselves). Perhaps then do not look at changes / tests for these runtimes for now: `mlserver-alibi-detect` and `mlserver-alibi-explain`.

note: I could not push to your fork so I created this [branch](https://github.com/sakoush/MLServer/commits/quickfix/py3.12/) as a reference to you. Feel free to use the commits on top of your changes as reference.

**Comment by sakoush** [comment]
@shivakrishnaah please have a look at the CI as there are numerous tests failing and they will need to be resolved as part of your change (exclude as discussed for now `alibi-explain` and `alibi-detect` tests). Please let me know if you have any questions.

**Review by sakoush — APPROVED** [review]
LGTM. Thanks for the initial work @shivakrishnaah I have done some extra work to get it over the line in the last few days. 

Some of the CI tasks still fail which I will investigate before merging.

**Comment by sakoush** [comment]
failing tests are all isolated in `test_lib_path`, we will look into it in a follow up PR.

**Pull request merged**